File: android/app/src/main/java/com/nexo/app/ui/admin/MemberDetailViewModel.kt
package com.nexo.app.ui.admin

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.nexo.app.data.repository.BackendRepository
import com.nexo.app.domain.model.ActivePlanItem
import com.nexo.app.domain.model.GymClass
import com.nexo.app.domain.model.GymMember
import com.nexo.app.domain.model.MembershipPlan
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Coordination logic for [MemberDetailSheet] — mirrors iOS's
 * `MemberDetailViewModel` (credit wallet grant/revoke, bookings with
 * staff-initiated cancel, remove-from-gym), kept out of the Composable
 * per CLAUDE.md's "business logic stays out of views" rule.
 */
class MemberDetailViewModel(
    private val repository: BackendRepository,
    private val gymId: String,
    val member: GymMember
) : ViewModel() {

    data class UiState(
        val isLoading: Boolean = true,
        val bookings: List<GymClass> = emptyList(),
        val activePlans: List<ActivePlanItem> = emptyList(),
        val availablePlans: List<MembershipPlan> = emptyList(),
        val didRemove: Boolean = false,
        val errorMessage: String? = null
    ) {
        val upcomingBookings: List<GymClass>
            get() = bookings.filter { it.startTimeMillis >= System.currentTimeMillis() }.sortedBy { it.startTimeMillis }

        val pastBookings: List<GymClass>
            get() = bookings.filter { it.startTimeMillis < System.currentTimeMillis() }.sortedByDescending { it.startTimeMillis }
    }

    private val _uiState = MutableStateFlow(UiState())
    val uiState: StateFlow<UiState> = _uiState.asStateFlow()

    init { load() }

    fun load() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            try {
                val bookings = repository.fetchMemberBookings(gymId, member.id)
                val activePlans = repository.fetchActivePlans(gymId, member.id)
                val availablePlans = repository.fetchMembershipPlans(gymId)
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    bookings = bookings,
                    activePlans = activePlans,
                    availablePlans = availablePlans
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(isLoading = false, errorMessage = "Error loading member: ${e.message}")
            }
        }
    }

    /** Staff-initiated cancel — mirrors iOS's `cancelBooking(gymId:classId:onBehalfOf:)`. */
    fun cancelBooking(classId: String) {
        viewModelScope.launch {
            try {
                repository.cancelBooking(gymId, classId, onBehalfOf = member.id)
                _uiState.value = _uiState.value.copy(bookings = _uiState.value.bookings.filter { it.id != classId })
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = "Failed to cancel booking: ${e.message}")
            }
        }
    }

    fun grantPlan(plan: MembershipPlan, customExpiresAtMillis: Long? = null) {
        viewModelScope.launch {
            try {
                repository.grantPlanToMember(gymId, member.id, plan, customExpiresAtMillis)
                _uiState.value = _uiState.value.copy(activePlans = repository.fetchActivePlans(gymId, member.id))
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = "Failed to grant plan: ${e.message}")
            }
        }
    }

    fun revokeActivePlan(activePlanId: String) {
        viewModelScope.launch {
            try {
                repository.revokeActivePlan(gymId, member.id, activePlanId)
                _uiState.value = _uiState.value.copy(activePlans = _uiState.value.activePlans.filter { it.id != activePlanId })
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = "Failed to revoke plan: ${e.message}")
            }
        }
    }

    /** Removes the member from the gym entirely. Sets [UiState.didRemove] on success so the screen knows to dismiss. */
    fun removeMember() {
        viewModelScope.launch {
            try {
                repository.removeMember(gymId, member.id)
                _uiState.value = _uiState.value.copy(didRemove = true)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(errorMessage = "Failed to remove member: ${e.message}")
            }
        }
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(errorMessage = null)
    }
}

File: ios/Nexo/Features/Members/MemberDetailViewModel.swift
//
//  MemberDetailViewModel.swift
//  Nexo
//

import Foundation

@MainActor
@Observable
final class MemberDetailViewModel {
    private let backend: BackendService
    let gymId: UUID
    let member: Member

    var bookings: [GymClass] = []
    var isLoading = false
    var errorMessage: String?
    /// The member's credit wallet — one item per granted plan component.
    var activePlans: [ActivePlanItem] = []
    /// The gym's plan templates, for the Grant Plan picker.
    var availablePlans: [MembershipPlan] = []

    var upcomingBookings: [GymClass] {
        bookings.filter { $0.startTime >= Date() }.sorted { $0.startTime < $1.startTime }
    }

    var pastBookings: [GymClass] {
        bookings.filter { $0.startTime < Date() }
    }

    init(gymId: UUID, member: Member, backend: BackendService? = nil) {
        self.gymId = gymId
        self.member = member
        self.backend = backend ?? FirebaseBackend.shared
    }

    func loadBookings() async {
        isLoading = true
        do {
            bookings = try await backend.fetchMemberBookings(gymId: gymId, userId: member.id)
        } catch {
            errorMessage = "Error loading bookings: \(error.localizedDescription)"
        }
        isLoading = false
    }

    func cancelBooking(_ gymClass: GymClass) async {
        do {
            try await backend.cancelBooking(gymId: gymId, classId: gymClass.id, onBehalfOf: member.id)
            bookings.removeAll { $0.id == gymClass.id }
        } catch {
            errorMessage = "Failed to cancel booking: \(error.localizedDescription)"
        }
    }

    // MARK: - Membership Plans & Credit Wallet

    func loadWallet() async {
        do {
            activePlans = try await backend.fetchActivePlans(gymId: gymId, userId: member.id)
        } catch {
            errorMessage = "Error loading wallet: \(error.localizedDescription)"
        }
    }

    func loadAvailablePlans() async {
        do {
            availablePlans = try await backend.fetchMembershipPlans(gymId: gymId)
        } catch {
            errorMessage = "Error loading plans: \(error.localizedDescription)"
        }
    }

    func grantPlan(_ plan: MembershipPlan, customExpiresAt: Date? = nil) async {
        do {
            try await backend.grantPlanToMember(gymId: gymId, userId: member.id, plan: plan, customExpiresAt: customExpiresAt)
            await loadWallet()
        } catch {
            errorMessage = "Failed to grant plan: \(error.localizedDescription)"
        }
    }

    func revokeActivePlan(_ item: ActivePlanItem) async {
        do {
            try await backend.revokeActivePlan(gymId: gymId, userId: member.id, activePlanId: item.id)
            activePlans.removeAll { $0.id == item.id }
        } catch {
            errorMessage = "Failed to revoke plan: \(error.localizedDescription)"
        }
    }

    /// Removes the member from the gym entirely. Returns whether it succeeded,
    /// so the view knows whether to dismiss.
    func removeMember() async -> Bool {
        do {
            try await backend.removeMember(gymId: gymId, userId: member.id)
            return true
        } catch {
            errorMessage = "Failed to remove member: \(error.localizedDescription)"
            return false
        }
    }
}
